gr-soapy/lib/block_impl.cc

```diff
@@ -1157,25 +1244,43 @@ std::string block_impl::read_uart(const std::string& which, long timeout_us) con
 void block_impl::cmd_handler_frequency(pmt::pmt_t val, size_t channel)
 {
     if (!(val->is_number() && !val->is_complex())) {
-        GR_LOG_WARN(d_logger, "soapy: freq must be float/int");
+        GR_LOG_ERROR(d_logger, "soapy: freq must be float/int");
         return;
     }
     set_frequency(channel, pmt::to_double(val));
 }
 
 void block_impl::cmd_handler_gain(pmt::pmt_t val, size_t channel)
 {
-    if (!(val->is_number() && !val->is_complex())) {
-        GR_LOG_WARN(d_logger, "soapy: gain must be float/int");
+    // For compatibility, accept either a numeric value for setting the
+    // overall gain or a dict with a gain element and element-specific
+    // value.
+    if (!(val->is_number() && !val->is_complex()) && !val->is_dict()) {
+        GR_LOG_ERROR(d_logger, "soapy: gain must be float/int or a dict");
         return;
     }
-    set_gain(channel, pmt::to_double(val));
+
+    if (val->is_dict()) {
+        if (!pmt::dict_has_key(val, CMD_GAIN_KEY)) {
+            GR_LOG_ERROR(d_logger, "soapy: gain dict must contain key \"gain\"");
+            return;
+        }
+
+        // Accept no name, falls back to default argument ""
+        const auto name =
+            pmt::symbol_to_string(pmt::dict_ref(val, CMD_NAME_KEY, pmt::mp("")));
```

Comment:
we're not expecting this to be a high-rate call, right? Because this should call `pmt::mp("")` every time, and that in turn hashes the empty string, does a hash lookup in the map of all known symbols, finds the previously register PMT symbol... for that then to be thrown away.

Reply:
I'll make a single instance up top with the keys and use that throughout.

Reply:
hey, that was an honest question about the rate. We shan't care if that happens 50 times a second!

You could do that static local or memer or if it's easier to read, split things to (attention, code that I'm not sure is prettier ahead)

```
auto& tmp = pmt::dict_ref(val, key, pmt::PMT_NIL);
std::string name = (tmp == PMT_NIL) ? "" : pmt::symbol_to_string(tmp);
```

or  so.